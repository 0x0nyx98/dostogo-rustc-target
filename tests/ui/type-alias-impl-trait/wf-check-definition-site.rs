// Regression test for #114572, We were inferring an ill-formed type:
//
// `Opaque<'a> = Static<&'a str>`, vs
// `Opaque<'a> = Static<&'static str>`.
#![feature(type_alias_impl_trait)]

struct Static<T: 'static>(T);

type OpaqueRet<'a> = impl Sized + 'a;
//~^ ERROR the type `&'a u8` does not fulfill the required lifetime
fn test_return<'a>(msg: Static<&'static u8>) -> OpaqueRet<'a> {
    msg
}

fn test_rpit<'a>(msg: Static<&'static u8>) -> impl Sized + 'a {
    //~^ ERROR the type `&'a u8` does not fulfill the required lifetime
    msg
}

type OpaqueAssign<'a> = impl Sized + 'a;
//~^ ERROR the type `&'a u8` does not fulfill the required lifetime
fn test_assign<'a>(msg: Static<&'static u8>) -> Option<OpaqueAssign<'a>> {
    let _: OpaqueAssign<'a> = msg;
    None
}

// `OpaqueRef<'a, T> = Ref<'a, T>`, vs
// `OpaqueRef<'a, T> = Ref<'static, T>`.
trait RefAt<'a>: 'a {}
struct Ref<'a, T: RefAt<'a>>(&'a T);
type OpaqueRef<'a, T: RefAt<'static>> = impl Sized + 'a;
//~^ ERROR mismatched types
fn test_trait<'a, T: RefAt<'static>>(msg: Ref<'static, T>) -> OpaqueRef<'a, T> {
    msg
}

fn main() {}
