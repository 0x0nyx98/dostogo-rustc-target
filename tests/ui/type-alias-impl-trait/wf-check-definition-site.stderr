error[E0477]: the type `&'a u8` does not fulfill the required lifetime
  --> $DIR/wf-check-definition-site.rs:9:22
   |
LL | type OpaqueRet<'a> = impl Sized + 'a;
   |                      ^^^^^^^^^^^^^^^
   |
note: type must satisfy the static lifetime as required by this binding
  --> $DIR/wf-check-definition-site.rs:7:18
   |
LL | struct Static<T: 'static>(T);
   |                  ^^^^^^^

error[E0477]: the type `&'a u8` does not fulfill the required lifetime
  --> $DIR/wf-check-definition-site.rs:15:47
   |
LL | fn test_rpit<'a>(msg: Static<&'static u8>) -> impl Sized + 'a {
   |                                               ^^^^^^^^^^^^^^^
   |
note: type must satisfy the static lifetime as required by this binding
  --> $DIR/wf-check-definition-site.rs:7:18
   |
LL | struct Static<T: 'static>(T);
   |                  ^^^^^^^

error[E0477]: the type `&'a u8` does not fulfill the required lifetime
  --> $DIR/wf-check-definition-site.rs:20:25
   |
LL | type OpaqueAssign<'a> = impl Sized + 'a;
   |                         ^^^^^^^^^^^^^^^
   |
note: type must satisfy the static lifetime as required by this binding
  --> $DIR/wf-check-definition-site.rs:7:18
   |
LL | struct Static<T: 'static>(T);
   |                  ^^^^^^^

error[E0308]: mismatched types
  --> $DIR/wf-check-definition-site.rs:31:41
   |
LL | type OpaqueRef<'a, T: RefAt<'static>> = impl Sized + 'a;
   |                                         ^^^^^^^^^^^^^^^ lifetime mismatch
   |
   = note: expected trait `RefAt<'a>`
              found trait `RefAt<'static>`
note: the lifetime `'a` as defined here...
  --> $DIR/wf-check-definition-site.rs:31:16
   |
LL | type OpaqueRef<'a, T: RefAt<'static>> = impl Sized + 'a;
   |                ^^
   = note: ...does not necessarily outlive the static lifetime

error: aborting due to 4 previous errors

Some errors have detailed explanations: E0308, E0477.
For more information about an error, try `rustc --explain E0308`.
